docs/elastic.rst

```diff
@@ -0,0 +1,358 @@
+
+2. Place all variables that need to be kept in sync between worker replicas (model parameters, optimizer state, epoch and batch numbers, etc.) into a ``hvd.elastic.State`` object.
+
+   Standard state implementations are provided for TensorFlow, Keras, and PyTorch.  However, it may be necessary in some cases to override
+   the base ``hvd.elastic.State`` object to handle broadcasting custom types.
+
+3. Periodically call ``state.commit()`` to backup a copy of your state in memory.
+
+   This is useful to prevent corrupted state in the event that a worker fails unexpectedly. For example, if training fails
+   in the middle of a parameter update, some gradient updates may have applied while others were still being allreduced.  When this
+   happens, a ``HorovodInternalError`` will be raised, and all parameters will be restored to the values at the time of the last commit.
+
+   Because commits can be expensive (as the model size increases), there is a tradeoff between the per-batch processing time
+   and how far the training process needs to rollback in the event of a failure.  For example, if you commit once every 10
+   batches, you reduce the amount of copying by a factor of 10. But if a failure occurs, you may need to redo up to 10
+   previously processed batches.
+
+   Elastic Horovod can avoid these rollbacks by performing what we call a *graceful removal* of a worker. If the driver
+   process discovers that a host has been made available or flagged for removal, it will push a notification to the workers.
+   The next time ``state.commit()`` or the more lightweight ``state.check_host_updates()`` is called, a ``HostsUpdatedInterrupt``
+   will be raised.  This event is handled similar to the ``HorovodInternalError``, except that parameter state will not be
+   restored to the last commit.
+
+   In general, if your hardware is generally reliable, and your orchestration system gives the driver ample warning
+   when a host is scheduled to be removed from the job, then you can safely call ``state.commit()`` on a reduced frequency,
+   and call ``state.check_host_updates()`` at the end of each batch instead.
+
+4. Register callbacks with the ``hvd.elastic.State`` object to respond to changes in the worker membership in the job.
+
+   For example, rescaling the learning rate with the new world size or repartitioning the dataset would commonly be done
+   through these callbacks.
+
+   Callbacks are called after Horovod has reinitialized, but before state is synchronized across the workers.
+
+The reset process following a ``HorovodInternalError`` (failure) or ``HostsUpdatedInterrupt`` (add/remove request) is as follows:
+
+1. Catch exception within the ``hvd.elastic.run`` decorator.
+2. Restore last committed state if ``HorovodInternalError`` was raised.
+3. Reinitialize Horovod context performing a new round of rendezvous.
+4. Synchronize state among the workers by broadcasting from the new worker-0.
```

Comment:
Is it possible that the `new worker-0` is the newly added worker whose weights are not synchronized with what has be obtained from the previous training?